### PR TITLE
Floating cursor cleanup

### LINKED ISSUE #30476: Support floating cursor double press on iOS
1 force press on the keyboard engages the floating cursor move mode.
A second force press leaves a selection anchor and then selects full words between the anchor and the current floating cursor position.

### PR DESCRIPTION
**Pull request by moffatman**
1. Floating cursor paint was being offset by the text field's scroll offset, but the floating cursor uses relative coordinates already. This resulted in the floating cursor being drawn in the wrong place when the text field was scrolled. (just a visual bug).
2. Floating cursor did not take into account selection affinity when calculating both start position and end position. This meant the behaviour either starting at or ending at the end of a line was not correct.
3. Don't change the selection when floating cursor is released if the new position would be contained within the current selection. Right now, only the baseOffset of the current selection is considered.
   * This won't change anything in practice on its own, but it is needed to avoid cancelling keyboard-select if selecting in the forwards direction.

Needed to prepare for https://github.com/flutter/engine/pull/36643

Needed to fix #30476

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 1]
```diff
@@ -2702,9 +2702,11 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
     final Offset finalPosition = renderEditable.getLocalRectForCaret(_lastTextPosition!).centerLeft - _floatingCursorOffset;
     if (_floatingCursorResetController!.isCompleted) {
       renderEditable.setFloatingCursor(FloatingCursorDragState.End, finalPosition, _lastTextPosition!);
-      if (_lastTextPosition!.offset != renderEditable.selection!.baseOffset) {
+      // Only change if new position is out of current selection range, as the
+      // selection may have been modified using the iOS keyboard selection gesture.
+      if (_lastTextPosition!.offset < renderEditable.selection!.baseOffset || _lastTextPosition!.offset > renderEditable.selection!.extentOffset) {
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/editable_text.dart:2707]: baseOffset can be greater than extentOffset. Use start/end instead.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 2]
```diff
@@ -2702,9 +2702,11 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
     final Offset finalPosition = renderEditable.getLocalRectForCaret(_lastTextPosition!).centerLeft - _floatingCursorOffset;
     if (_floatingCursorResetController!.isCompleted) {
       renderEditable.setFloatingCursor(FloatingCursorDragState.End, finalPosition, _lastTextPosition!);
-      if (_lastTextPosition!.offset != renderEditable.selection!.baseOffset) {
+      // Only change if new position is out of current selection range, as the
+      // selection may have been modified using the iOS keyboard selection gesture.
+      if (_lastTextPosition!.offset < renderEditable.selection!.start || _lastTextPosition!.offset >= renderEditable.selection!.end) {
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/editable_text.dart:2707]: Looks like this changes the behavior of the floating cursor when the initial selection is not empty if the floating cursor lands within the selected range?

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart:2707]: In a UIKit app the floating cursor removes the selection on start it seems

Comment (moffatman) [reply packages/flutter/lib/src/widgets/editable_text.dart:2707]: The selection is being cleared in flutter apps too by UIKit when floating cursor starts, so there shouldn't be any change.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 3]
```diff
@@ -2702,9 +2702,11 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
...
         // The cause is technically the force cursor, but the cause is listed as tap as the desired functionality is the same.
-        _handleSelectionChanged(TextSelection.collapsed(offset: _lastTextPosition!.offset), SelectionChangedCause.forcePress);
+        _handleSelectionChanged(TextSelection.collapsed(offset: _lastTextPosition!.offset, affinity: _lastTextPosition!.affinity), SelectionChangedCause.forcePress);
```
[the remark this answers is not in the document]
Comment (moffatman) [reply packages/flutter/lib/src/widgets/editable_text.dart:2709]: On line 2674 it's a position from the current selection. There isn't any `TextPosition.fromSelection`.

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 2 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart:2707]: Ah it already does that without the engine PR?

Comment (moffatman) [reply packages/flutter/lib/src/widgets/editable_text.dart:2707]: Yeah, I just checked in a flutter app built without the PR

**Review by LongCatIsLooong — APPROVED** [review]

**Pull request merged**